actions/rules/action_state_to_metastate.go

```diff
@@ -0,0 +1,459 @@
+	"github.com/fabric8-services/fabric8-wit/actions/change"
+	"github.com/fabric8-services/fabric8-wit/id"
+
+	errs "github.com/pkg/errors"
+	uuid "github.com/satori/go.uuid"
+
+	"github.com/fabric8-services/fabric8-wit/application"
+	"github.com/fabric8-services/fabric8-wit/workitem"
+)
+
+// ActionStateToMetaState implements the bidirectional mapping between state and column.
+type ActionStateToMetaState struct {
+	Db     application.DB
+	Ctx    context.Context
+	UserID *uuid.UUID
+}
+
+// make sure the rule is implementing the interface.
+var _ Action = ActionStateToMetaState{}
+
+func (act ActionStateToMetaState) contains(s []interface{}, e interface{}) bool {
+	for _, a := range s {
+		if a == e {
+			return true
+		}
+	}
+	return false
+}
+
+func (act ActionStateToMetaState) removeElement(s []interface{}, e interface{}) []interface{} {
+	for idx, a := range s {
+		if a == e {
+			copy(s[idx:], s[idx+1:])
+			s[len(s)-1] = nil
+			s = s[:len(s)-1]
+			// we don't return here as there may be multiple copies of e in s.
+		}
+	}
+	return s
+}
+
+// difference returns the differences between two slices. It returns two slices containing
+// the added and removed elements compared from old and new arguments.
+func (act ActionStateToMetaState) difference(old []interface{}, new []interface{}) ([]interface{}, []interface{}) {
+	var added []interface{}
+	var removed []interface{}
+	// added in slice2
+	for _, elem := range new {
+		if !act.contains(old, elem) {
+			added = append(added, elem)
+		}
+	}
+	// removed in slice2
+	for _, elem := range old {
+		if !act.contains(new, elem) {
+			removed = append(removed, elem)
+		}
+	}
+	return added, removed
+}
+
+func (act ActionStateToMetaState) loadWorkItemBoardsBySpaceID(spaceID uuid.UUID) ([]*workitem.Board, error) {
+	space, err := act.Db.Spaces().Load(act.Ctx, spaceID)
+	if err != nil {
+		return nil, errs.Wrap(err, "error loading space: "+err.Error())
+	}
+	boards, err := act.Db.Boards().List(act.Ctx, space.SpaceTemplateID)
+	if err != nil {
+		return nil, errs.Wrap(err, "error loading work item type: "+err.Error())
+	}
+	return boards, nil
+}
+
+func (act ActionStateToMetaState) loadWorkItemByID(id uuid.UUID) (*workitem.WorkItem, error) {
+	wi, err := act.Db.WorkItems().LoadByID(act.Ctx, id)
+	if err != nil {
+		return nil, errs.Wrap(err, "error loading work item: "+err.Error())
+	}
+	return wi, nil
+}
+
+func (act ActionStateToMetaState) storeWorkItem(workitem *workitem.WorkItem) (*workitem.WorkItem, error) {
+	err := application.Transactional(act.Db, func(appl application.Application) error {
+		var err error
+		workitem, _, err = appl.WorkItems().Save(act.Ctx, workitem.SpaceID, *workitem, *act.UserID)
+		if err != nil {
+			return errs.Wrap(err, "error updating work item")
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return workitem, nil
+}
+
+func (act ActionStateToMetaState) getValueListFromEnumField(wit *workitem.WorkItemType, fieldName string) ([]interface{}, error) {
```

Comment:
Why is `wit` a pointer? Let's not use pointers at unnecessary places.

Reply:
Fixed. 5cd9aef.